lib/vector-core/src/event/estimated_json_encoded_size_of.rs

```diff
@@ -83,17 +83,24 @@ impl EstimatedJsonEncodedSizeOf for Value {
     }
 }
 
-/// For performance reasons, strings aren't checked for the need for escape characters, nor for the
-/// need for UTF-8 replacement characters.
+/// For performance reasons, strings aren't checked for the need for UTF-8 replacement characters.
 ///
 /// This is the main reason why `EstimatedJsonEncodedSizeOf` is named as is, as most other types can
 /// be calculated exactly without a noticeable performance penalty.
 impl EstimatedJsonEncodedSizeOf for str {
     fn estimated_json_encoded_size_of(&self) -> JsonSize {
-        JsonSize::new(QUOTES_SIZE + self.len())
+        JsonSize::new(QUOTES_SIZE + self.len() + count_escapes(self.as_bytes()))
     }
 }
 
+// This is an imprecise estimate of the number of characters that will need to be escaped when
+// serializing to JSON. The goal is improve our overall accuracy in the case of relatively common
+// scenarios like strings of nested JSON, etc.
+fn count_escapes(input: &[u8]) -> usize {
+    memchr::memchr3_iter(b'"', b'\\', b'\t', input).count()
+        + memchr::memchr2_iter(b'\n', b'\r', input).count()
```

Comment:
Have you tried doing it in reverse? i.e. counting the number of times that `"\"\\\t\n\r".contains(i)` for each `i` in `input`?

Reply:
In an isolated benchmark, this is much faster for very small strings and much much slower for everything else.